src/proper.erl

```diff
@@ -1273,21 +1418,102 @@ get_rerun_result(#fail{}) ->
 get_rerun_result({error,_Reason} = ErrorResult) ->
     ErrorResult.
 
+-spec check_if_early_fail() -> 'ok'.
+check_if_early_fail() ->
+    Id = get('$property_id'),
+    receive
+        {worker_msg, {failed_test, From}, Id} ->
+            Passed = get('$tests_passed'),
+            From ! {worker_msg, {performed, Passed, Id}},
+            ok
+        after 0 -> ok
+    end.
+
+-spec update_tests_passed(non_neg_integer()) -> non_neg_integer().
+update_tests_passed(Passed) ->
+    case get('$tests_passed') of
+        undefined -> put('$tests_passed', Passed);
+        Passed2 -> put('$tests_passed', Passed + Passed2)
+    end.
+
+-spec perform(non_neg_integer(), test(), opts()) -> imm_result() | 'ok'.
 perform(NumTests, Test, Opts) ->
     perform(0, NumTests, ?MAX_TRIES_FACTOR * NumTests, Test, none, none, Opts).
 
+-spec perform(non_neg_integer(), pos_integer(), test(), opts()) -> imm_result() | 'ok'.
+perform(Passed, NumTests, Test, Opts) ->
+    Size = size_at_nth_test(Passed, Opts),
+    put('$size', Size),
+    perform(Passed, NumTests, 3 * ?MAX_TRIES_FACTOR * NumTests, Test, none, none, Opts).
```

Comment:
You should probably add a comment where this magic number `3` comes from... what is its role.